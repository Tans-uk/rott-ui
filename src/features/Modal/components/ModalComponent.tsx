import {isValidElement, useEffect, useMemo, useRef, type FC} from 'react'

import {InteractionManager, Platform, Modal as RNModal} from 'react-native'

import {Content} from '@features/Content'
import {Header, type HeaderProps} from '@features/Header'
import {Icon} from '@features/Icon'
import {Item} from '@features/Item'
import {
  Modal,
  ModalContentContainer,
  ModalStyles,
  PanResponderAnimation,
  type ModalProps,
} from '@features/Modal'
import {Pressable} from '@features/Pressable'
import {getHasDynamicIslandState, getHasNotchState, useSafeArea} from '@hooks'
import {addMatcherWait} from '@libs'
import {themeConfig} from '@providers'
import {colorFromVariant, display} from '@utils'

import {KeyboardStickyView} from 'react-native-keyboard-controller'
import Animated, {
  runOnJS,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated'

export const ModalComponent: FC<ModalProps> = ({
  id,
  animationType = 'fade',
  fullScreen,
  visible = false,

  header,
  closeButton,
  onClose,
  disableOutsideClick,

  height = fullScreen ? 100 : 0,
  backgroundColor = fullScreen
    ? themeConfig.colors['grey-900']
    : themeConfig.colors['neutral-alpha-100'],
  panResponderBackgroundColor = 'grey-800',
  /**
   * Custom Header element verildiği zaman tasarımda renk bozulması yaşamamak için bu değerin tanımlanması gerekir.
   *
   */
  headerBackgroundColor = 'grey-800',
  slideToClose,

  alignItemsCenter = undefined,
  justifyContentCenter = undefined,
  transparent,

  children,
  modals,
  modalContainerMarginBottom,
  sticksToKeyboard = false,
  ...props
}) => {
  const hasChildModal =
    modals?.length && modals?.length > 1 && id ? modals?.some((modal) => modal.id! > id) : false

  const {outsideTapAreaTestId, headerTestId, slideToCloseTestId} = {
    outsideTapAreaTestId: 'outside-tap-area-test-id',
    headerTestId: 'modal-header-test-id',
    slideToCloseTestId: 'slide-to-close-button-test-id',
  }

  const hasNotch = useAppSelector(getHasNotchState)
  const hasDynamicIsland = useAppSelector(getHasDynamicIslandState)
  const interactionRef = useRef<number>(undefined)

  const {bottom} = useSafeArea()

  const maxHeight = display.setHeightDevice(100)

  /** Height 100 verildiyse veya fullScreen ise kullanılan cihazı baz alır.
   * Eğer 0 ile 100 arası bir height verildiyse bu değer referans cihaza göre hesaplanır.
   * Örn, height: 50 -> referans cihaz yüksekliği 844 olduğundan 422 olarak hesaplanır.
   * Hesaplanan değer kullanılan cihazın yüksekliğinden fazla çıkarsa maksimum değer (%100) alınır.
   * Android'e özel olarak bottom inset eklenir.
   */
  const modalHeightByPercentage =
    fullScreen || height === 100
      ? maxHeight
      : Math.min(
          maxHeight,
          display.setHeight(height > 100 || height < 0 ? 100 : height) +
            (Platform.OS === 'android' ? bottom : 0) // ios'te sistem navigasyonu şeffaf ve boşluklu olduğu için bu düzeltmeye ihtiyaç duyulmaz
        )

  const canStickToKeyboard = !fullScreen && height < 50

  const translateY = useSharedValue(0)

  const animatedStyles = useAnimatedStyle(
    () => ({
      transform: [{translateY: translateY.value}],
    }),
    []
  )

  const closeAnimationDuration = 250

  const panResponder = useMemo(
    () =>
      PanResponderAnimation(
        modalHeightByPercentage,
        translateY,
        onClose ? () => onClose() : () => Modal.hideModal(id)
      ),
    [translateY, id, modalHeightByPercentage]
  )

  const closeModalAnimation = () => {
    if (!fullScreen && animationType === 'fade') {
      translateY.value = withTiming(modalHeightByPercentage, {duration: 250}, () => {
        !!onClose && runOnJS(onClose)()
        runOnJS(Modal.hideModal)(id)
      })
    } else {
      !!onClose && runOnJS(onClose)()
      runOnJS(Modal.hideModal)(id)
    }
  }
  useEffect(() => {
    if (visible) translateY.value = 0
    else {
      interactionRef.current = InteractionManager?.createInteractionHandle()
      addMatcherWait(
        () =>
          interactionRef.current &&
          InteractionManager.clearInteractionHandle(interactionRef.current),
        closeAnimationDuration
      )
    }
  }, [visible])

  useEffect(() => {
    return () => {
      interactionRef.current && InteractionManager.clearInteractionHandle(interactionRef.current)
    }
  }, [])

  return (
    <RNModal
      statusBarTranslucent // önemli (eski tip android cihazlarda edge-to-edge açık olsa dahi RNModal'a özel belirtmek gerekiyor)
      navigationBarTranslucent // önemli (eski tip android cihazlarda edge-to-edge açık olsa dahi RNModal'a özel belirtmek gerekiyor)
      transparent={!fullScreen}
      animationType={animationType}
      style={{height: modalHeightByPercentage}}
      visible={visible}
      onRequestClose={() => !disableOutsideClick && closeModalAnimation()}
      {...props}>
      {!fullScreen && (
        <Pressable
          testID={outsideTapAreaTestId}
          onPress={() => {
            if (!fullScreen && !!onClose && !disableOutsideClick) closeModalAnimation()
          }}
          style={ModalStyles().fadedBackgroundStyles}
        />
      )}
      <KeyboardStickyView enabled={canStickToKeyboard && sticksToKeyboard}>
        <Content noPadding backgroundColor={themeConfig.colors['neutral-alpha-700']} row>
          <Animated.View style={[ModalStyles().animatedViewStyles, animatedStyles]}>
            <ModalContentContainer
              height={modalHeightByPercentage}
              backgroundColor={transparent ? 'transparent' : backgroundColor}
              justifyContentCenter={justifyContentCenter}
              alignItemsCenter={alignItemsCenter}
              modalContainerMarginBottom={modalContainerMarginBottom}>
              {header && (
                <Item
                  size='full'
                  paddingTop={hasNotch || hasDynamicIsland ? (fullScreen ? 54 : 0) : 8}
                  backgroundColor={
                    headerBackgroundColor
                      ? colorFromVariant(headerBackgroundColor)
                      : fullScreen
                        ? themeConfig.colors['grey-800']
                        : themeConfig.colors['grey-900']
                  }
                  borderTopStartRadius={fullScreen ? 0 : 24}
                  borderTopEndRadius={fullScreen ? 0 : 24}
                  justifyContentCenter={!isValidElement(header)}
                  alignItemsCenter={!isValidElement(header)}

                  // TODO: FullScreen Modalda SideToClose Ozelligi var ise Header onPress methodlarini ezdigi icin bu ozellik kullanilmamaktadir. FullScreen Modal da sideToClose ozelligi isteniyorsa bu ozellik iyilestirilmeli ve daha sonra aktif edilmelidir.
                  // {...(fullScreen && slideToClose ? {...panResponder().panHandlers} : undefined)}
                >
                  {!fullScreen && slideToClose && (
                    <Item
                      size='full'
                      alignItemsCenter
                      justifyContentCenter
                      backgroundColor={
                        panResponderBackgroundColor
                          ? colorFromVariant(panResponderBackgroundColor)
                          : undefined
                      }
                      borderTopStartRadius={fullScreen ? 0 : 24}
                      borderTopEndRadius={fullScreen ? 0 : 24}
                      {...panResponder.panHandlers}>
                      <Icon
                        testID={slideToCloseTestId}
                        variant='neutral-grey-alpha-200'
                        width={40}
                        height={40}
                        name='LINE'
                      />
                    </Item>
                  )}

                  {isValidElement(header) && header}
                  {!isValidElement(header) && (
                    <Header
                      testID={headerTestId}
                      title={(header as HeaderProps)?.title}
                      subTitle={(header as HeaderProps)?.subTitle}
                      logo={(header as HeaderProps)?.logo}
                      paddingHorizontal={
                        (header as HeaderProps) ? (header as HeaderProps).paddingHorizontal : 0
                      }
                      rightIcon={
                        closeButton
                          ? {name: 'REMOVE_BIG', onPress: () => closeModalAnimation!()}
                          : (header as HeaderProps)?.rightIcon
                            ? (header as HeaderProps)?.rightIcon
                            : undefined
                      }
                      leftIcon={(header as HeaderProps)?.leftIcon}
                      {...(header as HeaderProps)}
                    />
                  )}
                </Item>
              )}

              {children}

              {hasChildModal && (
                <ModalComponent
                  {...modals?.filter((modal) => modal.id !== id)[0]}
                  modals={modals?.filter((modal) => modal.id !== id)}
                />
              )}
            </ModalContentContainer>
          </Animated.View>
        </Content>
      </KeyboardStickyView>
    </RNModal>
  )
}
