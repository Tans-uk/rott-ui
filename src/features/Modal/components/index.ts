export * from './ModalComponent'
export * from './ModalContainer'
