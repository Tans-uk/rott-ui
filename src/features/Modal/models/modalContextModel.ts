import {ThemeConfig} from '../../../models'
import type {ModalProps} from './modalProps'

export interface ModalContextModel {
  modals: ModalProps<ThemeConfig>[]
  showModal: (modalToRender: ModalProps<ThemeConfig>) => void
  updateModal: (modalToRender: ModalProps<ThemeConfig>, id?: number) => void
  hideModal: (id?: number) => void
  hasModalById: (id: number) => boolean
  hideAllModal: (
    predicate: (
      value: ModalProps<ThemeConfig>,
      index: number,
      array: ModalProps<ThemeConfig>[]
    ) => unknown
  ) => void
}
