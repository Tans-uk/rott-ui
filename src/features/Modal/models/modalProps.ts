import type {PropsWithChildren, ReactNode} from 'react'

import type {ModalProps as RNModalProps} from 'react-native'

import type {Variant} from '../../../models'
import type {HeaderProps} from '../../Header'

export interface ModalProps
  extends PropsWithChildren,
    Omit<RNModalProps, 'transparent' | 'animated' | 'id'> {
  id?: number
  fullScreen?: boolean

  header?: ReactNode | HeaderProps
  closeButton?: boolean
  onClose?: () => void
  disableOutsideClick?: boolean

  /**
   * Değer 0 ila 100 arasında olmalıdır.
   */
  height?: number
  backgroundColor?: string
  panResponderBackgroundColor?: Variant
  headerBackgroundColor?: Variant
  slideToClose?: boolean

  alignItemsCenter?: boolean
  justifyContentCenter?: boolean

  /**
   * transparent özelliği aktif edildiğinde item istendiği gibi sağa sola kaydırılabilir hale gelmektedir.
   *
   * InformationModal bu senaryoya ihtiyaç duymuş ve onun için geliştirilmiştir.
   * Bu özellik aktif edilmezse backgroundColor='transparent' verilse dahi arka plana tıklandığında modal kapanmamaktadır.
   * */
  transparent?: boolean
  modals?: ModalProps[]

  modalContainerMarginBottom?: number
  children?: any

  /** Modal açıkken soft klavye aktive edildiğinde modalın klavyeye yapışık halde sürüklenmesini sağlar.
   * Eğer modal belli bir uzunluktan yüksek ise bu prop dikkate alınmaz.
   */
  sticksToKeyboard?: boolean
}
