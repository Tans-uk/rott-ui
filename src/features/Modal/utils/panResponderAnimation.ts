import {PanResponder} from 'react-native'

import {Easing, runOnJS, withTiming, type SharedValue} from 'react-native-reanimated'

export const PanResponderAnimation = (
  modalHeight: number,
  translateY: SharedValue<number>,
  animationCloseCallBack: any
) =>
  PanResponder.create({
    onStartShouldSetPanResponder: () => true,
    onMoveShouldSetPanResponder: () => true,
    onPanResponderMove: (_, gestureState) => {
      if (gestureState.dy < 0) return
      if (gestureState.dy > 10 || gestureState.vy < 0) translateY.value = gestureState.dy
    },
    onPanResponderRelease: (_, gestureState) => {
      if (gestureState.dy > 100) {
        translateY.value = withTiming(modalHeight, {duration: 250}, () => {
          !!animationCloseCallBack && runOnJS(animationCloseCallBack)()
        })
      } else translateY.value = withTiming(0, {easing: Easing.elastic()})
    },
  })
