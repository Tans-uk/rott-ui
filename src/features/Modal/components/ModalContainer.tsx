/* eslint-disable react-native/no-inline-styles */
import {type FC} from 'react'

import {ScrollView, type GestureResponderEvent} from 'react-native'

import {type ModalProps} from '../models'

import {Content} from '@features/Content'
import {Pressable} from '@features/Pressable'
import {useSafeArea} from '@hooks'

interface ModalContainerProps extends ModalProps {
  onPress?: (event: GestureResponderEvent) => void
}

export const ModalContentContainer: FC<ModalContainerProps> = ({
  transparent,
  height,
  backgroundColor,
  fullScreen,
  justifyContentCenter,
  alignItemsCenter,
  modalContainerMarginBottom,
  onPress,
  children,
}) => {
  const {bottom} = useSafeArea()

  if (transparent) {
    return (
      <Content
        height={height}
        noPadding
        flex={1}
        backgroundColor={backgroundColor}
        borderTopStartRadius={fullScreen ? 0 : 24}
        borderTopEndRadius={fullScreen ? 0 : 24}
        justifyContentCenter={justifyContentCenter}
        alignItemsCenter={alignItemsCenter}
        marginBottom={modalContainerMarginBottom}>
        <ScrollView
          keyboardShouldPersistTaps='handled'
          showsVerticalScrollIndicator={false}
          showsHorizontalScrollIndicator={false}
          scrollEnabled={false}
          contentContainerStyle={{
            flex: fullScreen ? 0 : 1,
            paddingBottom: bottom,
            justifyContent: justifyContentCenter ? 'center' : undefined,
            alignItems: alignItemsCenter ? 'center' : undefined,
          }}>
          {children}
        </ScrollView>
      </Content>
    )
  } else {
    return (
      <Pressable
        onPress={onPress}
        height={height}
        backgroundColor={backgroundColor}
        borderTopStartRadius={fullScreen ? 0 : 24}
        borderTopEndRadius={fullScreen ? 0 : 24}
        justifyContentCenter={justifyContentCenter}
        alignItemsCenter={alignItemsCenter}
        marginBottom={modalContainerMarginBottom}>
        <ScrollView
          keyboardShouldPersistTaps='handled'
          showsVerticalScrollIndicator={false}
          showsHorizontalScrollIndicator={false}
          scrollEnabled={false}
          contentContainerStyle={{
            flex: fullScreen ? 0 : 1,
            paddingBottom: bottom,
            justifyContent: justifyContentCenter ? 'center' : undefined,
            alignItems: alignItemsCenter ? 'center' : undefined,
          }}>
          {children}
        </ScrollView>
      </Pressable>
    )
  }
}
