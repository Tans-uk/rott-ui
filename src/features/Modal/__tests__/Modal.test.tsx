import React from 'react'

import {fireEvent, render, waitFor} from '../../../__tests__/utils/testUtils'
import {translator} from '../../../libs'
import {ThemeConfig} from '../../../models'
import {Content} from '../../Content'
import {Item} from '../../Item'
import {Label} from '../../Label'
import {ModalComponent} from '../components'
import type {ModalProps} from '../models'

describe('Modal -> Custom Component', () => {
  const onCloseMock = jest.fn()

  const testId = {
    modalTestId: 'modal-test-id',
    headerTestId: 'modal-header-test-id',
    headerCloseButtonTestId: 'header-right-pressable-test-id',
    headerCloseIconTestId: 'header-right-icon-test-id',
    outsideTapAreaTestId: 'outside-tap-area-test-id',
    slideToCloseTestId: 'slide-to-close-button-test-id',
  }

  const dummyDataFullScreen: ModalProps<ThemeConfig> = {
    testID: testId.modalTestId,
    header: {
      title: translator('TEST'),
      logo: 'PTTBANK_BLACK_COLORED',
    },
    slideToClose: true,
    visible: true,
    fullScreen: true,
    closeButton: true,
    onClose: onCloseMock,
  }

  const dummyData: ModalProps<ThemeConfig> = {
    testID: testId.modalTestId,
    header: {
      title: translator('TEST'),
      logo: 'PTTBANK_BLACK_COLORED',
    },
    visible: true,
    height: 50,
    fullScreen: false,
    slideToClose: true,
    closeButton: true,
    onClose: onCloseMock,
  }

  beforeAll(() => {
    // TODO: useState mock example
    const setStateMock = jest.fn()
    const useStateMock: any = (useState: any) => [useState, setStateMock]
    jest.spyOn(React, 'useState').mockImplementation(useStateMock)
  })

  it('modal ilk renderlandiginda snapshot ile eşleşmeli', () => {
    const renderedModal = render(
      <ModalComponent {...dummyData}>
        <Item>
          <Label>{translator('TEST')}</Label>
        </Item>
      </ModalComponent>
    )

    expect(renderedModal).toMatchSnapshot()
  })

  it('modal visible değilken ekranda gözükmemeli', () => {
    const {queryByText} = render(
      <ModalComponent {...dummyData} visible={false}>
        <Item>
          <Label>{translator('TEST.WITH.PARAM', {testText: 'test'})}</Label>
        </Item>
      </ModalComponent>
    )

    const testLabelByText = queryByText(translator('TEST.WITH.PARAM', {testText: 'test'}))

    expect(testLabelByText).not.toBeOnTheScreen()
  })

  it('custom header verildiğinde ekranda custom header olmalı', () => {
    const {headerTestId} = testId

    const {getByTestId} = render(
      <ModalComponent
        {...dummyData}
        header={
          <Content testID={headerTestId}>
            <Label>{translator('TEST')}</Label>
          </Content>
        }>
        <Item>
          <Label>{translator('TEST.WITH.PARAM', {testText: 'test'})}</Label>
        </Item>
      </ModalComponent>
    )

    const modalHeader = getByTestId(headerTestId)

    expect(modalHeader).toBeOnTheScreen()
  })

  it('modal kapatma butonu headerda olmalı', () => {
    const {headerCloseIconTestId} = testId
    const {getByTestId} = render(<ModalComponent {...dummyData} />)

    const headerCloseButton = getByTestId(headerCloseIconTestId)

    expect(headerCloseButton).toBeOnTheScreen()
  })

  it('modal header child elementi ekranda olmalı', () => {
    const {getByTestId} = render(
      <ModalComponent
        {...dummyData}
        header={{
          title: translator('TEST'),
          logo: 'PTTBANK_BLACK_COLORED',
          children: <Label testID={'header-children-test-id'}>{translator('TEST')}</Label>,
        }}>
        <Item>
          <Label>{translator('TEST.WITH.PARAM', {testText: 'test'})}</Label>
        </Item>
      </ModalComponent>
    )

    const headerChildren = getByTestId('header-children-test-id')

    expect(headerChildren).toBeOnTheScreen()
  })

  it('modal kapatma butonuna tıklandığında modal kapanmalı', async () => {
    const onClosePressableMock = jest.fn()
    const {headerCloseButtonTestId} = testId
    const {getByTestId} = render(<ModalComponent {...dummyData} onClose={onClosePressableMock} />)

    const headerCloseButton = getByTestId(headerCloseButtonTestId)
    fireEvent.press(headerCloseButton)

    await waitFor(() => expect(onClosePressableMock).toHaveBeenCalledTimes(1))
  })

  it('modal tam ekran değilken container dışına tıklanırsa modal kapanmalı', async () => {
    const onClosePressableMock = jest.fn()
    const {outsideTapAreaTestId} = testId
    const {getByTestId} = render(<ModalComponent {...dummyData} onClose={onClosePressableMock} />)

    const outsideContainer = getByTestId(outsideTapAreaTestId)
    fireEvent.press(outsideContainer)

    await waitFor(() => expect(onClosePressableMock).toHaveBeenCalledTimes(1))
  })

  it('panResponder headerda olmalı', () => {
    const {slideToCloseTestId} = testId
    const {getByTestId} = render(<ModalComponent {...dummyData} />)

    const panResponder = getByTestId(slideToCloseTestId)

    expect(panResponder).toBeOnTheScreen()
  })

  it('panResponder headerda olmamalı', () => {
    const {slideToCloseTestId} = testId
    const {queryByTestId} = render(<ModalComponent {...dummyData} slideToClose={false} />)

    const panResponder = queryByTestId(slideToCloseTestId)

    expect(panResponder).toBeNull()
  })

  it('panResponder header dışında olmalı', () => {
    const {slideToCloseTestId, headerTestId} = testId
    const {getByTestId} = render(<ModalComponent {...dummyData} />)

    const panResponder = getByTestId(slideToCloseTestId)

    // TODO: İç elementleri de bir katman olarak çağırıyor.
    expect(panResponder.parent?.parent?.parent).not.toHaveProp('testID', headerTestId)
  })

  it('modal fullscreen modunda panResponder ekranda olmamalı', () => {
    const {slideToCloseTestId} = testId
    const {queryByTestId} = render(<ModalComponent {...dummyDataFullScreen} />)

    const panResponder = queryByTestId(slideToCloseTestId)

    expect(panResponder).not.toBeOnTheScreen()
  })

  it('modal verilen children elementi ekranda göstermeli', () => {
    const {getByText} = render(
      <ModalComponent fullScreen visible>
        <Label>{translator('TEST')}</Label>
      </ModalComponent>
    )

    const testLabelByText = getByText(translator('TEST'))

    expect(testLabelByText).toBeOnTheScreen()
  })
})
