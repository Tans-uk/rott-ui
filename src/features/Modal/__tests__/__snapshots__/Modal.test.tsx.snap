// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal -> Custom Component modal ilk renderlandiginda snapshot ile eşleşmeli 1`] = `
<View
  flex={1}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    navigationBarTranslucent={true}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    style={
      {
        "height": 422,
      }
    }
    testID="modal-test-id"
    transparent={true}
    visible={true}
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      animated={false}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignContent": "center",
          "backgroundColor": "rgba(34, 63, 70, 0.75)",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
      testID="outside-tap-area-test-id"
    />
    <View
      enabled={false}
    >
      <View
        style={
          {
            "backgroundColor": "rgba(34, 63, 70, 0.75)",
            "flex": undefined,
            "flexDirection": "row",
            "paddingBottom": 0,
            "paddingHorizontal": 0,
          }
        }
      >
        <View
          style={
            [
              {
                "elevation": 5,
                "flex": 1,
                "flexDirection": "row",
                "shadowColor": "rgba(6, 14, 16, 0.30)",
                "shadowOffset": {
                  "height": -8,
                  "width": 0,
                },
                "shadowOpacity": 1,
                "shadowRadius": 35,
              },
              {
                "transform": [
                  {
                    "translateY": 0,
                  },
                ],
              },
            ]
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            animated={false}
            backgroundColor="rgba(34, 63, 70, 0.10)"
            borderTopEndRadius={24}
            borderTopStartRadius={24}
            collapsable={false}
            focusable={true}
            height={422}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "rgba(34, 63, 70, 0.10)",
                "flex": 1,
                "flexDirection": "column",
                "height": 812,
                "maxHeight": 812,
              }
            }
            testID="pressable-test-id"
          >
            <RCTScrollView
              contentContainerStyle={
                {
                  "alignItems": undefined,
                  "flex": 1,
                  "justifyContent": undefined,
                  "paddingBottom": 0,
                }
              }
              keyboardShouldPersistTaps="handled"
              scrollEnabled={false}
              showsHorizontalScrollIndicator={false}
              showsVerticalScrollIndicator={false}
            >
              <View>
                <View
                  alignItemsCenter={true}
                  backgroundColor="rgba(61, 88, 94, 1)"
                  borderTopEndRadius={24}
                  borderTopStartRadius={24}
                  justifyContentCenter={true}
                  paddingTop={8}
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "rgba(61, 88, 94, 1)",
                      "flex": undefined,
                      "flexDirection": "column",
                      "justifyContent": "center",
                      "paddingTop": 16,
                      "width": "100%",
                    }
                  }
                >
                  <View
                    alignItemsCenter={true}
                    backgroundColor="rgba(61, 88, 94, 1)"
                    borderTopEndRadius={24}
                    borderTopStartRadius={24}
                    justifyContentCenter={true}
                    onMoveShouldSetResponder={[Function]}
                    onMoveShouldSetResponderCapture={[Function]}
                    onResponderEnd={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderReject={[Function]}
                    onResponderRelease={[Function]}
                    onResponderStart={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    onStartShouldSetResponderCapture={[Function]}
                    style={
                      {
                        "alignItems": "center",
                        "backgroundColor": "rgba(61, 88, 94, 1)",
                        "flex": undefined,
                        "flexDirection": "column",
                        "justifyContent": "center",
                        "width": "100%",
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "flex": undefined,
                          "flexDirection": "column",
                        }
                      }
                      testID="slide-to-close-button-test-id"
                    >
                      <Text
                        fill="rgba(234, 239, 240, 0.15)"
                        height={77}
                        strokeWidth={0}
                        testID="mock-icon"
                        width={77}
                      >
                        MockIcon
                      </Text>
                    </View>
                  </View>
                  <View
                    minHeight={56}
                    paddingVertical={0}
                    style={
                      {
                        "alignItems": "center",
                        "flex": undefined,
                        "flexDirection": "column",
                        "justifyContent": "center",
                        "minHeight": 108,
                        "paddingBottom": 0,
                        "paddingHorizontal": 46,
                        "paddingVertical": 0,
                      }
                    }
                    testID="modal-header-test-id"
                  >
                    <View
                      height={56}
                      style={
                        {
                          "flex": undefined,
                          "flexDirection": "row",
                          "height": 108,
                          "maxHeight": 108,
                        }
                      }
                    >
                      <View
                        flexShrink={1}
                        height={56}
                        justifyContentCenter={true}
                        minWidth={56}
                        relative={true}
                        style={
                          {
                            "flex": undefined,
                            "flexDirection": "column",
                            "flexShrink": 1,
                            "height": 108,
                            "justifyContent": "center",
                            "maxHeight": 108,
                            "minWidth": 108,
                            "position": "relative",
                          }
                        }
                      >
                        <View
                          accessibilityState={
                            {
                              "busy": undefined,
                              "checked": undefined,
                              "disabled": undefined,
                              "expanded": undefined,
                              "selected": undefined,
                            }
                          }
                          accessibilityValue={
                            {
                              "max": undefined,
                              "min": undefined,
                              "now": undefined,
                              "text": undefined,
                            }
                          }
                          accessible={true}
                          alignItemsCenter={false}
                          alignItemsFlexStart={true}
                          animated={false}
                          collapsable={false}
                          focusable={true}
                          height={56}
                          justifyContentCenter={true}
                          minWidth={56}
                          onBlur={[Function]}
                          onClick={[Function]}
                          onFocus={[Function]}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderRelease={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            {
                              "alignItems": "flex-start",
                              "flex": 1,
                              "flexDirection": "column",
                              "height": 108,
                              "justifyContent": "center",
                              "maxHeight": 108,
                              "minWidth": 108,
                              "width": 108,
                            }
                          }
                          testID="header-left-pressable-test-id"
                          width={56}
                        >
                          <View
                            style={
                              {
                                "flex": undefined,
                                "flexDirection": "column",
                              }
                            }
                            testID="header-left-icon-test-id"
                          >
                            <Text
                              fill="rgba(255, 255, 255, 1)"
                              height={46}
                              strokeWidth={0}
                              testID="mock-icon"
                              width={46}
                            >
                              MockIcon
                            </Text>
                          </View>
                        </View>
                      </View>
                      <View
                        alignItemsCenter={true}
                        flex={1}
                        height={56}
                        justifyContentCenter={true}
                        minWidth={56}
                        relative={true}
                        style={
                          {
                            "alignItems": "center",
                            "flex": 1,
                            "flexDirection": "column",
                            "height": 108,
                            "justifyContent": "center",
                            "maxHeight": 108,
                            "minWidth": 108,
                            "position": "relative",
                          }
                        }
                      >
                        <Image
                          height={24}
                          source={
                            {
                              "testUri": "../../../src/assets/images/logo/pttbank-BlackColored.png",
                            }
                          }
                          style={
                            {
                              "height": 46,
                              "maxHeight": 46,
                              "overflow": "hidden",
                              "resizeMode": "contain",
                              "width": 244,
                            }
                          }
                          testID="header-logo-test-id"
                          width={127}
                        />
                      </View>
                      <View
                        alignItemsFlexEnd={true}
                        flexShrink={1}
                        height={56}
                        justifyContentCenter={true}
                        minWidth={56}
                        relative={true}
                        style={
                          {
                            "alignItems": "flex-end",
                            "flex": undefined,
                            "flexDirection": "column",
                            "flexShrink": 1,
                            "height": 108,
                            "justifyContent": "center",
                            "maxHeight": 108,
                            "minWidth": 108,
                            "position": "relative",
                          }
                        }
                      >
                        <View
                          accessibilityState={
                            {
                              "busy": undefined,
                              "checked": undefined,
                              "disabled": undefined,
                              "expanded": undefined,
                              "selected": undefined,
                            }
                          }
                          accessibilityValue={
                            {
                              "max": undefined,
                              "min": undefined,
                              "now": undefined,
                              "text": undefined,
                            }
                          }
                          accessible={true}
                          alignItemsCenter={false}
                          alignItemsFlexEnd={true}
                          alignItemsFlexStart={true}
                          animated={false}
                          collapsable={false}
                          focusable={true}
                          height={56}
                          justifyContentCenter={true}
                          minWidth={56}
                          name="REMOVE_BIG"
                          onBlur={[Function]}
                          onClick={[Function]}
                          onFocus={[Function]}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderRelease={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            {
                              "alignItems": "flex-end",
                              "flex": 1,
                              "flexDirection": "column",
                              "height": 108,
                              "justifyContent": "center",
                              "maxHeight": 108,
                              "minWidth": 108,
                            }
                          }
                          testID="header-right-pressable-test-id"
                        >
                          <View
                            onPress={[Function]}
                            style={
                              {
                                "flex": undefined,
                                "flexDirection": "column",
                              }
                            }
                            testID="header-right-icon-test-id"
                          >
                            <Text
                              fill="rgba(255, 255, 255, 1)"
                              height={46}
                              strokeWidth={0}
                              testID="mock-icon"
                              width={46}
                            >
                              MockIcon
                            </Text>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
                <View
                  style={
                    {
                      "flex": undefined,
                      "flexDirection": "column",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "color": "rgba(17, 17, 17, 1)",
                        "fontFamily": "Markpro-Medium",
                        "fontSize": 14,
                        "fontWeight": undefined,
                        "textAlign": "auto",
                      }
                    }
                  >
                    ASD
                  </Text>
                </View>
              </View>
            </RCTScrollView>
          </View>
        </View>
      </View>
    </View>
  </Modal>
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "flexDirection": "column-reverse",
          "justifyContent": "flex-start",
          "top": 30,
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "bottom": 40,
          "flexDirection": "column",
          "justifyContent": "flex-end",
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>
</View>
`;
